eclair-core/src/main/scala/fr/acinq/eclair/channel/TxPublisher.scala

```diff
@@ -0,0 +1,413 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.channel
+
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.{ByteVector32, ByteVector64, Crypto, Satoshi, Script, Transaction, TxOut}
+import fr.acinq.eclair.NodeParams
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient
+import fr.acinq.eclair.blockchain.bitcoind.rpc.ExtendedBitcoinClient.FundTransactionOptions
+import fr.acinq.eclair.blockchain.fee.FeeratePerKw
+import fr.acinq.eclair.blockchain.{CurrentBlockCount, WatchConfirmed, WatchEventConfirmed}
+import fr.acinq.eclair.transactions.Transactions._
+import fr.acinq.eclair.transactions.{Scripts, Transactions}
+import fr.acinq.eclair.wire.protocol.UpdateFulfillHtlc
+
+import java.util.concurrent.Executors
+import scala.collection.immutable.SortedMap
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+ * Created by t-bast on 25/03/2021.
+ */
+
+/**
+ * This actor ensures its parent channel's on-chain transactions confirm in a timely manner.
+ * It sets the fees, tracks confirmation progress and bumps fees if necessary.
+ */
+object TxPublisher {
+
+  // @formatter:off
+  sealed trait Command
+  sealed trait PublishTx extends Command {
+    def tx: Transaction
+  }
+  /**  Publish a fully signed transaction without modifying it. */
+  case class PublishRawTx(tx: Transaction) extends PublishTx
+  /**
+   * Publish an unsigned transaction. Once (csv and cltv) delays have been satisfied, the tx publisher will set the fees,
+   * sign the transaction and broadcast it.
+   */
+  case class SignAndPublishTx(txInfo: TransactionWithInputInfo, commitments: Commitments) extends PublishTx {
+    override def tx: Transaction = txInfo.tx
+  }
+  case class WrappedCurrentBlockCount(currentBlockCount: Long) extends Command
+  case class ParentTxConfirmed(childTx: PublishTx, parentTxId: ByteVector32) extends Command
+  private case class PublishNextBlock(p: PublishTx) extends Command
+  // @formatter:on
+
+  // NOTE: we use a single thread to publish transactions so that it preserves order.
+  // CHANGING THIS WILL RESULT IN CONCURRENCY ISSUES WHILE PUBLISHING PARENT AND CHILD TXS!
+  val singleThreadExecutionContext = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())
+
+  def apply(nodeParams: NodeParams, watcher: akka.actor.ActorRef, client: ExtendedBitcoinClient): Behavior[Command] =
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[CurrentBlockCount](cbc => WrappedCurrentBlockCount(cbc.blockCount)))
+      new TxPublisher(nodeParams, watcher, client, context)(singleThreadExecutionContext).run(SortedMap.empty, Map.empty)
+    }
```

Comment:
One of the great advantages of making this publisher a child of the channel is that we can now contextualize all logs with an MDC, which is awesome for debugging. Here it would be `nodeId` and `channelId`.

Here is an example:
https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/ChannelRelay.scala#L53-L59

Reply:
It's unfortunately not that simple due to the complexity of the `Channel` initialization...
We don't have the `channelId` available easily.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1749/commits/7239391f41248f24cc9d1d9933713ba76e7bdda9
It seems to work fine!